### DISCUSSION TITLE
New plugin for Corsair Commander PRO

### DISCUSSION THREAD

**Discussion by ghost** [Rem0o/FanControl.Releases#602] [Show and tell]
Hi everybody,

No doubt there will be a few people that found this fantastic little gem thanks to JayzTwoCents, I did and am glad that I did!  Great job Rem0o!

So, I found that FanControl was not able to report or control the speed of my fans connected via the Corsair Commander PRO - not a great surprise but it irked me, and getting rid of iCUE is always nice ;)

Long story short I've thrown a plugin together that allows FanControl to do just that!  Please keep in mind the following:

- This is the first time I have used GitHub to host any code - so my repo is looking a bit drab (to say the least).  Please be gentle.
- Most importantly - I only have 5 fans connected out of the possible 6.  It **should** work for all 6 - but we all know the difference between should and does ;)

Repo is linked below, please let me know if you have any issues.

[https://github.com/iJacks1980/FanControl.CommanderPRO](https://github.com/iJacks1980/FanControl.CommanderPRO)

**Comment by Jonathanese** [comment]
Sweet! I'll have to check this out.

FanControl + Commander Pro would be a phenomenal combo.

**Reply by Jonathanese** [reply]
OK I got it up and running. Had to completely uninstall iCue because it REALLY messes with the sensors.

Next step will be to find a way to integrate the fans on the H115i.

**Reply by rizlah** [reply]
yeah icue is a pain, it likes to take control of everything and not release it.  I need to keep icue running though for my RGB, same issue i had with SyncRGB, i needed to keep icue running to control my fans, if i can get this working then i could use syncrgb for my rgb and fancontrol for the fans, however with icue being the pain in the ass it is i dont want to remove it until i know this will work :)

**Reply by Jonathanese** [reply]
@iJacks1980 I'm looking at the code and it looks like it wouldn't be too hard to add the H115i stuff myself, perhaps in a fork.

I'd be curious where you got the information on how to pull the Corsair info.

**Reply by ghost** [reply]
Funnily enough Rem0o asked me the same question about the protocol :)  I honestly can't remember, I was dabbling with trying to get rid of iCUE a while back and making my own "Fan Controller", back when NZXT and others couldn't read Ryzen CPU temperatures - couldn't get it to work, of course NOW I know it's because I had iCUE installed ;)  Can't remember the source of the info though - I know it didn't include any details regarding RGB or Temp data though.

Feel free to fork if you want, or do a PR - have to be honest not sure how those work - I've been TFS/Azure DevOps for years and never touched GIT.  My plugin will probably become obsolete once LibreHardwareMonitor is working with the Commander PRO's.

**Reply by ghost** [reply]
Hi rizlah,

I'm not familiar with SyncRGB - does that use the iCUE software to send RGB commands?  If so you might want to look at (OpenRGB)[https://openrgb.org/] - that supports the Commander PRO and it doesn't need iCUE.  Alas it's not something that I can test as my RGB died, not sure if it's the controller or the fans.

**Reply by Jonathanese** [reply]
Lol yeah I'm not sure about PRs either. I've been working with small businesses for years doing one-off projects. My latest job is the first time I've had multiple hands in the same code. Might do some digging. But your plugin is working pretty good so far. A real life-saver!

**Comment by rizlah** [comment]
hey, i have downloaded the 5 files and put them in the plugin directory, restarted the fancontrol software but it did not detect any of my corsair fan controllers

any suggestions

**Reply by Jonathanese** [reply]
I had to work with the files a bit. Turned out I had to have JUST the Fancontrol.commanderpro.dll (or whatever the file is actually called) in the Plugins folder with none of the other files.

**Reply by rizlah** [reply]
i still do not see the plugin being loaded, even just putting that file in the plugin folder

**Reply by AtA3301** [reply]
> i still do not see the plugin being loaded, even just putting that file in the plugin folder

You did restart FanControl after adding the files...? Just checking ;-)

**Reply by rizlah** [reply]
Yeah sure did

**Reply by Fr0stX76** [reply]
Did you try "refresh sensor detection" from the top right menu (3 dots menu)?

**Reply by rizlah** [reply]
Yeah but can try that Again. Not sure if I did it with just the one file in the plug-in folder

**Reply by ghost** [reply]
I know you resolved this specific issue rizlah, but in case anybody stumbles upon this the file was blocked by Windows by default - and can only be unblocked by going through a right click, properties, check the unblock this file box and pressing ok.

I have a love/hate relationship with this feature in Windows.  On the one hand it makes sense to try and protect users - on the other it's a PITA when random files won't load but do so silently!

**Reply by aca-bobo** [reply]
Download link is 404. I need this file please

**Reply by Impact123** [reply]
@aca-bobo 
See https://github.com/Rem0o/FanControl.Releases/discussions/1226

**Comment by AtA3301** [comment]
> Hi everybody,
> 
> No doubt there will be a few people that found this fantastic little gem thanks to JayzTwoCents, I did and am glad that I did! Great job Rem0o!
> 
> So, I found that FanControl was not able to report or control the speed of my fans connected via the Corsair Commander PRO - not a great surprise but it irked me, and getting rid of iCUE is always nice ;)
> 
> Long story short I've thrown a plugin together that allows FanControl to do just that! Please keep in mind the following:
> 
> * This is the first time I have used GitHub to host any code - so my repo is looking a bit drab (to say the least).  Please be gentle.
> * Most importantly - I only have 5 fans connected out of the possible 6.  It **should** work for all 6 - but we all know the difference between should and does ;)
> 
> Repo is linked below, please let me know if you have any issues.
> 
> https://github.com/iJacks1980/FanControl.CommanderPRO

Amazing, within a few hours too. 🍻

**Comment by eddwatts** [comment]
Just got this working, my reported rpm is all over the place, but the speed control is working as wanted! nice to have all fans working in 1 place and being able to mix the fan curves is what I wanted.

Amazing job on the plugin, esp given the speed!

is it worth adding the dll download so people dont have to compile it themselvs?

**Reply by rizlah** [reply]
I thought the dll was already there. I downloaded one from here but can't get it to work.

**Reply by eddwatts** [reply]
> I thought the dll was already there. I downloaded one from here but can't get it to work.

This is used during the build, load "FanControl Plugins.sln" into visual Studio, install anything needed, right click "FanControl.CommanderPro" on the right and then build, this creates a "FanControl.CommanderPro.dll" in a bin folder (mine was 12KB in size) this then gets copied to your plugins folder.

**Reply by Fr0stX76** [reply]
Or... just download FanControl.CommanderPro.pdb and FanControl.CommanderPro.dll from https://github.com/iJacks1980/FanControl.CommanderPRO/releases

While fancontrol is closed, put both in the plugin folder. start fancontrol, activate the plugin in the sensor menu (top right 3 dots) Then refresh sensors

If this doesnt work, it could be a firmware incompatibility between the plugin and your com. pro. raise a ticket in the plugin page

**Reply by rizlah** [reply]
did as you suggested, except there is no option to enable the plugin

![image](https://user-images.githubusercontent.com/56230853/164896845-70f3df4d-03df-44ff-a006-05ff7aad106e.png)

**Reply by rizlah** [reply]
this is the error i get in the log

File name: 'file:///E:\Utilities\FanControl\Plugins\FanControl.CommanderPro.dll' ---> System.NotSupportedException: An attempt was made to load an assembly from a network location which would have caused the assembly to be sandboxed in previous versions of the .NET Framework. This release of the .NET Framework does not enable CAS policy by default, so this load may be dangerous. If this load is not intended to sandbox the assembly, please enable the loadFromRemoteSources switch. See http://go.microsoft.com/fwlink/?LinkId=155569 for more information.
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadFrom(String assemblyFile, Evidence securityEvidence, Byte[] hashValue, AssemblyHashAlgorithm hashAlgorithm, Boolean forIntrospection, Boolean suppressSecurityChecks, StackCrawlMark& stackMark)
   at System.Reflection.Assembly.LoadFrom(String assemblyFile)
   at FanControl.Domain.PluginService.LoadAssembly(FileInfo dll)

**Reply by Rem0o** [reply]
Unlock the DLL => Right click properties => Checkbox at the bottom

**Reply by rizlah** [reply]
FIXED :)

**Reply by rizlah** [reply]
when i say fixed, i can see them, but i cant seem to control them, when i put them to a curve they stay at 0% but they are spinning, they occasionally flash that they are being controlled and show the correct %age but then go back to 0% and it still shows no plugin loaded

![image](https://user-images.githubusercontent.com/56230853/164908723-2450572e-54be-43cb-8586-3488db2ffbf4.png)

**Reply by Fr0stX76** [reply]
Did you do a full shutdown since closing Icue? I ask because you might need to reet the com pro

**Reply by rizlah** [reply]
i havent closed icue as i need it for rgb and other stuff

**Reply by Fr0stX76** [reply]
Well... that will probably conflict with both loaded. icue is known not to play well with any other software

**Reply by rizlah** [reply]
yeah i did wonder, i know icue can be a pain and not like to release things to be controlled by other software, appears i will not be able to use this for anything other than my exhaust fan and GPU, still great software though

**Reply by Jonathanese** [reply]
Basically it fixed everything when I completely uninstalled iCue. I know that isn't much of an option for you though.

It's one of the services. You may have to prevent one of the corsair services from running. I'm not sure which one or if it will also disable lighting control, though.

**Reply by ghost** [reply]
eddwatts, thanks for the kind words.

With regards to the RGB have you tried looking at the [OpenRGB](https://openrgb.org/) project?  It supports the Commander PRO - my RGB controller died (or the fan RGB's have died, they are very wonky) so I don't use iCUE at all any more.  Not sure if the UI is as smart as iCUE - in that iCUE lets you select the specific fans to set the number of RGB's to be controlled, but might be worth a look?

**Reply by rizlah** [reply]
you could always try signalRGB, best RGB software i have found, works with all RGB devices

**Comment by Fr0stX76** [comment]
Is E your windows partition? If C is then try moving the whole thing from E drive to C: Program files

**Reply by rizlah** [reply]
moved it to c:\program files, same issue

**Reply by Fr0stX76** [reply]
http://www.itlec.com/2016/05/systemnotsupportedexception-attempt-was.html

**Reply by irlrobins** [reply]
I'm having similar difficulties in getting this plugin to run. I've downloaded the FanControl.CommanderPro.dll to plugin folder, made sure my FanControl application folder was under c:\program files and that there was no block on the .dll. The plugin is detected:

![image](https://user-images.githubusercontent.com/104267065/164912047-50b82a83-7737-497a-9e4c-31fa58101fd7.png)
but non of the 6 fans connected to the Commander respond to changing the controls manually:
![image](https://user-images.githubusercontent.com/104267065/164912087-68948ec5-6151-45bf-994d-3c36fd606097.png)

I've uninstalled iCUE, so it's not running at the same time as FanControl either.

**Reply by Rem0o** [reply]
@irlrobins  Please open issues and such in the plugin's repo.  There might be a command missing being sent to the control when enabled/disabled.

**Reply by ghost** [reply]
Hi irlrobins,

Please do raise an issue on my repo - but in the meantime take a look in the main Fan Control folder.  My plugin has it's own error log file "CommanderPRO.err.log" - and this should be appended to whenever there is an exception.  If you have this file attach it to an issue and I will look into it.

**Comment by ghost** [comment]
Hi guys,

Well, wasn't expecting that - I went to bed and woke up to a fair few emails ;)  

So I know about the conflict with iCUE - nothing that can be done about that, it's the way the hardware works.  However I'll work on detecting iCUE and see if I can at least get detection working (or at least fail gracefully).  I'll have to force iCUE back on myself for that ;)

I'll clean up the release as well - turns out it had more files than were needed (which makes sense), but I did throw this together at half two in the morning.

I'll also look into getting the temp probes working - if I can find mine, I never connected them as I had no need.

If anybody has any other issues please let me know by raising an issue on my  [repo](https://github.com/iJacks1980/FanControl.CommanderPRO/issues) - I'm sure Rem0o has enough work just keeping up with Fan Control requests :)

I'll also update the repo readme with know issues/fixes shortly.

**Reply by Jonathanese** [reply]
Yeah I don't expect much in the way of working with iCue running. I think just a Corsair problem.

**Comment by ghost** [comment]
Hi all,

I've just posted a [pre-release version of 1.0.1](https://github.com/iJacks1980/FanControl.CommanderPRO/releases/tag/Beta) that adds support for reading the temperature of connected temp probes to the Commander PRO.

It also has a disabled plugin - which I hope to have working in a few days time - for the Commander CORE, so when you see both Commander PRO and CORE listed in the Fan Control plugins please don't expect any CORE devices to show up...yet ;)

**Comment by Revvet** [comment]
I have a commander core XT so I will be watching for the updated version.  While I wait, I do have a question.  Is it possible to just use ICUE to control RGB and not fans by removing my 4 Corsair LL case fan inputs from the Commander Core XT and my 2 Corsair ML AIO fans to my Corsair H115I Platnum AIO inputs and run them all into the MOBO fan headers?  Thanks for any help

**Reply by irlrobins** [reply]
Yes, this is an option. Fans would be controlled directly by MB headers. You would need splitters obviously to connect more than one fan to each header.

**Reply by ghost** [reply]
My brand new H115i should be with me in a few hours, so once I have it installed and working (and redone all my cable management!) I'll get back to working the Commander CORE code.

Revvet, I can't be certain that the XT and non XT versions are the same - so I will mark the XT as being experimental, I might need your input to help test compatibility for it.

**Reply by Revvet** [reply]
Not a problem iJacks.  I will leave my setup as-is for now and try out the XT version when it is ready.

**Comment by Rem0o** [comment]
@iJacks1980 featured your plugin in the https://github.com/Rem0o/FanControl.Releases/blob/master/README.md

**Reply by Revvet** [reply]
I assume this is still for the Commander Pro and not for the Commander Core XT yet?

**Reply by ghost** [reply]
Yes, still working on the core version - it's like wrestling a bull, I thought I had it then it threw me for 6.

I actually have the code reading the Firmware, connected fans and fan speeds (and I think I will make the first release a read only) - but it's not reliable yet.

**Comment by hst12** [comment]
Working nicely for me. My system is actually much less noisier now than with iCue for the same cooling effect.

One request if I may: Would it be possible to read the temperature probes?

**Reply by ghost** [reply]
Temp probes are already supported in the [1.0.1 Pre Release](https://github.com/iJacks1980/FanControl.CommanderPRO/releases/tag/Beta) for the Commander PRO.

**Reply by hst12** [reply]
Thanks for the help. The plugin was blocked by Windows all along.

**Comment by CovertRockhound** [comment]
Just a heads up, things get all screwy when using this and also using AIDA64 with "Corsair Link sensor support" enabled. Speeds bounce around and misreport and stuff similarly to having FanControl and iCUE running at the same time. 

This is fairly unfortunate for me as I wanted to use FanControl and have a AIDA64 sensor panel displaying fan speeds at the same time. It seems this wont work together. And using iCUE and AIDA64 at the same time also results in this problem. I kind of have no options now that im aware of besides just picking which one I value the most, and not using the least valuable one.

**Reply by Fr0stX76** [reply]
Sadly the issue is an hardware limitation due to corsair not wanting 3rd party to access their echosystem. Just look at their support forum, they are limit arrogant to anyone who just mention that using 3rd party software will cause issue.

Bottom line is they have interesting products but once you are in the echosystem, unless you want to make compromises it`s either you stay in it or you leave it.

**Comment by Revvet** [comment]
Yeah, I have used Corsair support many times and if 3rd party is mentioned, they will not help.  I think the easiest thing for me to do is remove all fans from the Commander Core XT and move them to Mobo headers and control them with Fan Control.  I would use ICUE only for the RGB control (which is working flawlessly).  I hope that is a viable option.  Can anyone comment on that since ICUE doesn't always play well with other open system monitoring/controlling applications?

**Reply by Fr0stX76** [reply]
From the plugin page: 

"Known Issues
If Corsair iCUE is installed you will get random data at best, nothing at worst. Sadly the way that the hardware works means things have to be one or the other :( If you need iCUE for the RGB I can recommend [OpenRGB](https://openrgb.org/)"

**Reply by Revvet** [reply]
Yes, I'm sorry, I had read that before but I thought it meant using Corsair for fan control and that if I removed all the fans from the core XT I might avoid the problem.  I guess I can try and see how it goes.  4 pin splitters are cheap so what the heck.  Otherwise, I will hope I can eventually use Fan Control using the plug-in Ijacks1980 is developing

**Reply by Revvet** [reply]
I also haven't looked at OpenRGB before but anything non-Corsair trying to run Corsair products worries me concerning functionality.  As we know, Corsair doesn't play well with others

**Reply by ghost** [reply]
In my experience there is nothing wrong with the various 3rd party utilities - but only if they are used on there own.  As OpenRGB (and others like SyncRGB) doesn't touch the fan speed data they should, in theory, work well with fan monitoring/control.

As an example Liquidctl does a great job in monitoring and controlling the pumps and fans of some Corsair products - but only if iCUE is not around.  Sadly unless Corsair decides to open up (and the requests for Linux support have been around for years with no positive response from them) there will always be issues.

I've managed to get fan data read from the Commander CORE device - as long as iCUE is nowhere to be seen - but even without it's still problematic.  Fan Control does seem to talk to the Commander CORE - indirectly - and this then locks my code out.  Thankfully we both use the same HID library so I am looking at implementing the various events that get fired to allow multiple programs to share device access - getting there...I think I might age a few years in a very short time though ;)

**Reply by Jonathanese** [reply]
@iJacks1980 Yeah there's really only so much you can do. Corsair didn't seem to write their driver or perhaps firmware with any intention of multiple connections to the same device. A potential solution would be to create an interface service that would handle communication with the hardware, and then have third party applications talk to this service rather than trying to talk to the device directly.

That actually sounds like another interesting project to take on. Maybe have it hide the original device and create a secondary virtual device that communicates with the service which iCue could then detect and use instead. Corsair already seems to have a set of services they use for control and data collection, but of course it is all walled off to third parties. Been a while since I've done any reverse engineering and that was for very simple microcontroller programs in college.

**Reply by Revvet** [reply]
I took a look at OpenRGB and that looks very interesting.  My biggest problem with RGB is I don't have enough Mobo headers for my case fans and would rather not lose some of the sequential effects ICUE can perform so I would need something like I have now with the Core XT.  Also, I have a mixture of addressable (4 pin) and 3 pin RGB fans, plus I have Corsair RGB RAM, a Corsair AIO Cooler, and an Asus GPU.  Will OpenRGB handle those RGBs??  If so, I'm on board with it.
